File: fe/src/component/Sector.js
import React from 'react';

const Sector = ({sectors, temp}) => {
    let count = ""
    return (
        <>
            {sectors.map(m => {
                if (m.parentId != null) {
                    count = temp + "\u00a0\u00a0"
                }

                return (
                    <>
                        <option value={m.id}>{`${count}  ${m.name}`}</option>
                        {m.sectors && <Sector temp={count} sectors={m.sectors}/>}
                    </>
                );
            })}
        </>
    );
}

export default Sector

File: be/src/main/java/com/example/sectors/entity/Sector.java
package com.example.sectors.entity;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;
import lombok.experimental.Accessors;

import javax.persistence.CascadeType;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import javax.validation.constraints.NotBlank;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@ToString
@Accessors(chain = true)
@Table(name = "sector")
public class Sector {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    private String name;

    @ToString.Exclude
    @OneToMany(mappedBy = "parentSector", cascade = CascadeType.ALL)
    private List<Sector> sectors = new ArrayList<>();

    @ManyToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "parent_id")
    private Sector parentSector;

}
